BlockChainRunner.java

```diff
@@ -0,0 +1,354 @@
+                        .build()
+                );
+            }
+        });
+        transactionPool.removeAll(block.getTransactions());
+
+        blockChain.add(block);
+    }
+
+    private long getCurrentTimestamp() {
+        return new Timestamp(System.currentTimeMillis()).getTime();
+    }
+
+    private void generatePendingTransactions() throws NoSuchAlgorithmException {
+        if (totalTransactionsPooled >= TRANSACTION_POOL_SIZE) return;
+        int userCount = users.size();
+
+        Signature sign = Signature.getInstance("SHA256withDSA");
+
+        unspentOutputMap.keySet().forEach(userKey -> {
+            List<UnspentOutput> userUnspentOutputs = new ArrayList<>(unspentOutputMap.get(userKey));
+            int balance = calculateUserBalance(userKey);
+            if (balance < 2) return;
+
+            int sumToSpend = ThreadLocalRandom.current().nextInt(1, balance);
+
+            List<TransactionInput> inputs = new ArrayList<>();
+            int sumAlreadySpend = 0;
+            User currentUser = pkUserMap.get(userKey);
+
+            Iterator<UnspentOutput> outputIterator = userUnspentOutputs.iterator();
+            while (sumAlreadySpend < sumToSpend && outputIterator.hasNext()) {
+                UnspentOutput output = outputIterator.next();
+                sumAlreadySpend += getUnspentOutputValue(output);
+
+                inputs.add(
+                    TransactionInput.builder()
+                        .index(output.getOutputIndex())
+                        .txId(output.getTxId())
+                        .signedTxId(currentUser.sign(sign, output.getTxId().getBytes()))
+                        .build()
+                );
+            }
+
+            List<TransactionOutput> outputs = new ArrayList<>();
+            int destUserIndex = ThreadLocalRandom.current().nextInt(0, userCount);
+            User destUser = users.get(destUserIndex);
+            outputs.add(
+                TransactionOutput.builder().value(sumToSpend).destPublicKey(destUser.getPublicKey()).build()
+            );
+            if (sumAlreadySpend > sumToSpend) {
+                outputs.add(
+                    TransactionOutput.builder().value(sumAlreadySpend - sumToSpend).destPublicKey(userKey).build()
+                );
+            }
+
+            pendingTransactions.add(new Transaction(inputs, outputs));
+        });
+    }
+
+    private void addPendingTransactionsToPool()
+        throws NoSuchAlgorithmException, SignatureException, InvalidKeyException
+    {
+        Signature sign = Signature.getInstance("SHA256withDSA");
+
+        for (Transaction pendingTransaction : pendingTransactions) {
+            boolean isValid = true;
+            List<Pair<PublicKey, UnspentOutput>> mappingsToRemove = new ArrayList<>();
+
+            // Validate transaction here ↓
+            int transactionInputValue = 0;
+
+            if (!pendingTransaction.getTxId().equals(pendingTransaction.computeTransactionId())) {
```

Comment:
The clause itself is hard to understand as there is no explanation of it, further more it only provides an INFO log message if the clause is passed.

Would be advisable to either rethink if the log itself is needed, or reformat it so it would be more understandable.

Reply:
Addressed